kubernetes/istio/istio/src/main/java/org/arquillian/cube/istio/impl/IstioResourcesApplier.java

```diff
@@ -1,55 +1,100 @@
+import java.util.logging.Logger;
 import me.snowdrop.istio.client.IstioClient;
 import org.arquillian.cube.istio.api.IstioResource;
 import org.arquillian.cube.kubernetes.impl.resolver.ResourceResolver;
+import org.arquillian.cube.kubernetes.impl.resources.KubernetesResourcesApplier;
 import org.arquillian.cube.kubernetes.impl.utils.RunnerExpressionParser;
 import org.jboss.arquillian.core.api.annotation.Observes;
 import org.jboss.arquillian.test.spi.TestClass;
+import org.jboss.arquillian.test.spi.event.suite.After;
 import org.jboss.arquillian.test.spi.event.suite.AfterClass;
+import org.jboss.arquillian.test.spi.event.suite.Before;
 import org.jboss.arquillian.test.spi.event.suite.BeforeClass;
 
-import java.io.BufferedInputStream;
-import java.io.IOException;
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.List;
-
 public class IstioResourcesApplier {
 
-    private List<HasMetadata> createdIstioResources = new ArrayList<>();
+    private static final Logger log = Logger.getLogger(KubernetesResourcesApplier.class.getName());
 
-    public void applyIstioResources(@Observes(precedence = -20) BeforeClass beforeClass, final IstioClient istioClient) {
+    private Map<String, List<me.snowdrop.istio.api.model.IstioResource>> resourcesMap = new ConcurrentHashMap<>();
 
+    public void applyIstioResourcesAtClassScope(@Observes(precedence = -20) BeforeClass beforeClass, final IstioClient istioClient) {
         final TestClass testClass = beforeClass.getTestClass();
 
-        Arrays.stream(findAnnotations(testClass))
+        log.info(String.format("Creating environment for %s", testClass.getName()));
+
+        createResources(createResourceKey(testClass), istioClient, findAnnotations(testClass));
+    }
+
+    public void applyIstioResourcesAtMethodScope(@Observes(precedence = -20) Before beforeMethod, final IstioClient istioClient) {
+        final TestClass testClass = beforeMethod.getTestClass();
+        final Method testMethod = beforeMethod.getTestMethod();
+
+        log.info(String.format("Creating environment for %s method %s", testClass.getName(), testMethod.getName()));
+
+        createResources(createResourceKey(testMethod), istioClient, findAnnotations(testMethod));
+    }
+
+    public void removeIstioResourcesAtClassScope(@Observes(precedence = 20) AfterClass afterClass, final IstioClient istioClient) {
+        final TestClass testClass = afterClass.getTestClass();
+
+        log.info(String.format("Deleting environment for %s", testClass.getName()));
+
+        deleteResources(createResourceKey(testClass), istioClient);
+    }
+
+    public void removeIstioResourcesAtMethodScope(@Observes(precedence = 20) After afterMethod, final IstioClient istioClient) {
+        final TestClass testClass = afterMethod.getTestClass();
+        final Method testMethod = afterMethod.getTestMethod();
+
+        log.info(String.format("Deleting environment for %s method %s", testClass.getName(), testMethod.getName()));
```

Comment:
Maybe instead of `environment` we could say Istio resource? to not confuse with kubernetes environment (aka namespace)